quarkus:dev fails with .m2/repo in path with spaces on all platforms

Download https://github.com/quarkusio/quarkus-quickstarts/tree/master/getting-started
run via
mvn compile quarkus:dev
throws
java.lang.NoClassDefFoundError: org/jboss/logging/Logger

on Windows10 with both Java8 and Java11:
Apache Maven 3.6.0 (97c98ec64a1fdfee7767ce5ffb20918da4f719f3; 2018-10-24T20:41:47+02:00)
Maven home: D:\Development\Maven\bin\..
Java version: 11.0.2, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk-11.0.2
Default locale: de_DE, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

exception:

Listening for transport dt_socket at address: 5005
Exception in thread "main" java.lang.NoClassDefFoundError: org/jboss/logging/Logger
        at io.quarkus.dev.DevModeMain.<clinit>(DevModeMain.java:39)
Caused by: java.lang.ClassNotFoundException: org.jboss.logging.Logger
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:583)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
        ... 1 more
[INFO] ------------------------------------------------------------------------

> [INFO] --- quarkus-maven-plugin:999-SNAPSHOT:dev (default-cli) @ getting-started ---
[INFO] Using servlet resources C:\Users\tandraschko\Downloads\getting-started\src\main\resources\META-INF\resources
Listening for transport dt_socket at address: 5005
Exception in thread "main" java.lang.NoClassDefFoundError: org/jboss/logging/Logger
        at io.quarkus.dev.DevModeMain.<clinit>(DevModeMain.java:39)
Caused by: java.lang.ClassNotFoundException: org.jboss.logging.Logger
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:338)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 1 more

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -23,6 +23,7 @@ import java.net.InetAddress;
 import java.net.Socket;
 import java.net.URI;
 import java.net.URL;
+import java.net.URLDecoder;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Arrays;

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -20,7 +20,6 @@ import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.net.InetAddress;
-import java.net.MalformedURLException;
 import java.net.Socket;
 import java.net.URI;
 import java.net.URL;
@@ -114,6 +113,9 @@ public class DevMojo extends AbstractMojo {
 
     @Parameter(defaultValue = "${session}")
     private MavenSession session;
+
+    @Parameter(defaultValue = "TRUE")
+    private boolean deleteDevJar;
 
     @Component
     private MavenVersionEnforcer mavenVersionEnforcer;
@@ -251,11 +253,12 @@ public class DevMojo extends AbstractMojo {
                 String jarPath = classFile.getPath().substring(0, classFile.getPath().lastIndexOf('!'));
                 if (jarPath.startsWith("file:"))
                     jarPath = jarPath.substring(5);
-                path = new File(jarPath);
+                // The resource will be URL encoded, so decode is so when addToClassPaths is called the encoding is correct
+                path = new File(URLDecoder.decode(jarPath, StandardCharsets.UTF_8.name()));
             } else if (classFile.getProtocol().equals("file")) {
                 String filePath = classFile.getPath().substring(0,
                         classFile.getPath().lastIndexOf(DevModeMain.class.getName().replace('.', '/')));
-                path = new File(filePath);
+                path = new File(URLDecoder.decode(filePath, StandardCharsets.UTF_8.name()));
             } else {
                 throw new MojoFailureException("Unsupported DevModeMain artifact URL:" + classFile);
             }
@@ -265,7 +268,11 @@ public class DevMojo extends AbstractMojo {
 
             File tempFile = new File(buildDir, project.getArtifactId() + "-dev.jar");
             tempFile.delete();
-            tempFile.deleteOnExit();
+            // Only delete the -dev.jar on exit if requested
+            if (deleteDevJar) {
+                tempFile.deleteOnExit();
+            }
+            getLog().debug("Executable jar: " + tempFile.getAbsolutePath());
 
             try (ZipOutputStream out = new ZipOutputStream(new FileOutputStream(tempFile))) {
                 out.putNextEntry(new ZipEntry("META-INF/"));
@@ -298,6 +305,8 @@ public class DevMojo extends AbstractMojo {
             args.add(outputDirectory.getAbsolutePath());
             args.add(wiringClassesDirectory.getAbsolutePath());
             args.add(new File(buildDir, "transformer-cache").getAbsolutePath());
+            // Display the launch command line in debug mode
+            getLog().debug("Launching JVM with comand line: " + args.toString());
             ProcessBuilder pb = new ProcessBuilder(args.toArray(new String[0]));
             pb.redirectError(ProcessBuilder.Redirect.INHERIT);
             pb.redirectOutput(ProcessBuilder.Redirect.INHERIT);
@@ -380,10 +389,10 @@ public class DevMojo extends AbstractMojo {
         return java;
     }
 
-    private void addToClassPaths(StringBuilder classPathManifest, StringBuilder classPath, File file)
-            throws MalformedURLException {
+    private void addToClassPaths(StringBuilder classPathManifest, StringBuilder classPath, File file) {
         URI uri = file.toPath().toAbsolutePath().toUri();
-        classPathManifest.append(uri.getRawPath());
+        String path = uri.getRawPath();
+        classPathManifest.append(path);
         classPath.append(uri.toString());
         if (file.isDirectory()) {
             classPathManifest.append("/");

